patches/api/0421-Paper-Item-API.patch

```diff
@@ -0,0 +1,582 @@
++ */
++public interface TropicalFishVariant {
++
++    /**
++     * Gets the pattern color
++     * @return color
++     */
++    @NotNull DyeColor patternColor();
++
++    /**
++     * Gets the body color
++     * @return color
++     */
++    @NotNull DyeColor bodyColor();
++
++    /**
++     * Gets the type
++     * @return type
++     */
++    @NotNull TropicalFish.Pattern type();
++}
+diff --git a/src/main/java/io/papermc/paper/property/MutablePropertyView.java b/src/main/java/io/papermc/paper/property/MutablePropertyView.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..d73a2276a3e56905e9595f85f5532f06f8dc583d
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/property/MutablePropertyView.java
+@@ -0,0 +1,56 @@
++package io.papermc.paper.property;
++
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++public interface MutablePropertyView extends PropertyView, Cloneable {
++
++    /**
++     * Unsets a property key.
++     *
++     * @param key key
++     */
++    void unset(@NotNull PropertyKey<?> key);
++
++    /**
++     * Sets the value of the given property key.
++     *
++     * @param key   key
++     * @param value value
++     * @param <T>   type
++     */
++    <T> void set(@NotNull PropertyKey<T> key, @Nullable T value);
++
++    /**
++     * Creates a new property key. TODO: I don't like this
++     *
++     * @param key
++     * @param <T>
++     * @return
++     */
++    @NotNull
++    <T> T create(@NotNull PropertyKey<T> key);
++
++    /**
++     * TODO: I don't like this
++     *
++     * @param key
++     * @param <T>
++     * @return
++     */
++    @NotNull
++    <T> T getOrCreate(@NotNull PropertyKey<T> key);
++
++    /**
++     * Clones the given property view.
++     *
++     * @return cloned value
++     */
++    @NotNull
++    MutablePropertyView clone();
++
++    /**
++     * Clears the property view.
++     */
++    void clear();
++}
+diff --git a/src/main/java/io/papermc/paper/property/PropertyKey.java b/src/main/java/io/papermc/paper/property/PropertyKey.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..4d02e565476a04cdccf78d700a14aa45586f7233
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/property/PropertyKey.java
+@@ -0,0 +1,72 @@
++package io.papermc.paper.property;
++
++import io.leangen.geantyref.TypeToken;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++import java.lang.reflect.Type;
++
++/**
++ * Represents a key that holds a property.
++ *
++ * @param <T> type
++ */
++@ApiStatus.Experimental
++public interface PropertyKey<T> {
++
++    /**
++     * Creates a property key with the provided key and class.
++     *
++     * @param key   string key
++     * @param token class type
++     * @param <T>   type
++     * @return property key
++     */
++    @NotNull
++    static <T> PropertyKey<T> of(@NotNull String key, @NotNull Class<T> token) {
```

Comment:
Still hard disagree that this should be exposed in any way to the API. We have the PDC for storing custom information

Reply:
Yeah no this won’t be public.